File: ios/Sources/ARYPush/Internal/PermissionManager.swift
import Foundation
import UIKit
import UserNotifications

/// Reads and requests notification authorization.
///
/// Two separate things decide whether a notification is delivered on iOS, and confusing them is
/// the usual source of "the token is fine but nothing appears":
///
/// * the user's authorization decision, which can only be asked for once;
/// * whether the app has actually registered with APNs, which is a separate call and which the
///   OS will not do for an unauthorized app.
///
/// This type owns the first. Registration is triggered by ``TokenManager``.
final class PermissionManager {

    private let storage: StorageManager
    private let center: UNUserNotificationCenter

    init(storage: StorageManager, center: UNUserNotificationCenter = .current()) {
        self.storage = storage
        self.center = center
    }

    /// Current normalised authorization state.
    func status() async -> PushPermissionStatus {
        let settings = await center.notificationSettings()
        return PushPermissionStatus.from(settings.authorizationStatus)
    }

    /// Callback form of ``status()``, for call sites that are not already async.
    func status(_ completion: @escaping (PushPermissionStatus) -> Void) {
        center.getNotificationSettings { settings in
            completion(PushPermissionStatus.from(settings.authorizationStatus))
        }
    }

    /// Whether the system prompt has already been shown.
    ///
    /// iOS answers `notDetermined` only until the first prompt, so this is mostly a convenience;
    /// it is recorded anyway so that the SDK never shows a prompt it knows will not appear.
    var hasBeenAsked: Bool {
        get { storage.bool(StorageManager.Keys.permissionRequested, default: false) }
        set { storage.set(StorageManager.Keys.permissionRequested, newValue) }
    }

    /// Requests authorization, presenting the system prompt when one is still possible.
    ///
    /// `completion` is invoked on the main queue exactly once, whatever the user does.
    func request(
        options: UNAuthorizationOptions,
        completion: @escaping (PushPermissionStatus) -> Void
    ) {
        center.getNotificationSettings { [weak self] settings in
            guard let self else { return }

            guard settings.authorizationStatus == .notDetermined else {
                // Already granted, or already denied: iOS will not present a second prompt, and
                // asking again silently does nothing. Report the truth instead.
                let current = PushPermissionStatus.from(settings.authorizationStatus)
                PushLogger.info("requestPermission(): no prompt is possible, status is \(current)")
                DispatchQueue.main.async { completion(current) }
                return
            }

            self.hasBeenAsked = true
            self.center.requestAuthorization(options: options) { _, error in
                if let error {
                    PushLogger.error("Authorization request failed: \(error)")
                }
                // The status is re-read rather than inferred from the granted flag, so
                // provisional and ephemeral authorization are reported accurately.
                self.status { status in
                    PushLogger.info("Permission result: \(status)")
                    DispatchQueue.main.async { completion(status) }
                }
            }
        }
    }

    /// Records the permission state and reports whether it changed.
    ///
    /// The backend needs to know when a device stops being reachable, and iOS never tells an
    /// application that the user switched notifications off in Settings, so this is sampled at
    /// every launch and after every permission request.
    func recordStatusChange(_ status: PushPermissionStatus) -> Bool {
        let previous = storage.string(StorageManager.Keys.lastPermissionState)
        guard previous != status.rawValue else { return false }
        storage.set(StorageManager.Keys.lastPermissionState, status.rawValue)
        PushLogger.info("Permission status: \(previous ?? "unknown") -> \(status.rawValue)")
        return true
    }

    /// Opens this application's page in Settings, the only route left once a user has denied.
    @MainActor
    func openSettings() {
        guard let url = URL(string: UIApplication.openSettingsURLString),
              UIApplication.shared.canOpenURL(url) else {
            PushLogger.warn("Could not open the application settings page")
            return
        }
        UIApplication.shared.open(url)
    }
}

File: android/sdk/src/main/kotlin/com/ary/push/internal/permission/PermissionManager.kt
package com.ary.push.internal.permission

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.provider.Settings
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import com.ary.push.internal.log.PushLogger
import com.ary.push.internal.storage.StorageManager
import com.ary.push.model.PushPermissionStatus

/**
 * Reads and requests the notification permission, hiding the Android 13 split.
 *
 * Two separate things decide whether a notification can be posted, and confusing them is the
 * usual source of "the token is fine but nothing appears":
 *
 *  * the `POST_NOTIFICATIONS` runtime permission, which only exists from API 33;
 *  * the application-level notification toggle, which exists on every version and which the user
 *    can switch off in Settings at any time.
 *
 * [status] reports the combination, so host code never has to branch on API level.
 */
internal class PermissionManager(
    context: Context,
    private val storage: StorageManager
) {

    private val appContext = context.applicationContext

    /** Current normalised permission state. */
    val status: PushPermissionStatus
        get() {
            val enabledAtAppLevel = runCatching {
                NotificationManagerCompat.from(appContext).areNotificationsEnabled()
            }.getOrDefault(true)

            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
                // Before API 33 there is nothing to grant: the app-level toggle is the whole story.
                return if (enabledAtAppLevel) {
                    PushPermissionStatus.GRANTED
                } else {
                    PushPermissionStatus.DENIED
                }
            }

            val granted = ContextCompat.checkSelfPermission(
                appContext,
                Manifest.permission.POST_NOTIFICATIONS
            ) == PackageManager.PERMISSION_GRANTED

            return when {
                granted && enabledAtAppLevel -> PushPermissionStatus.GRANTED
                granted -> PushPermissionStatus.DENIED // Granted, then switched off in Settings.
                hasBeenAsked -> PushPermissionStatus.DENIED
                else -> PushPermissionStatus.NOT_DETERMINED
            }
        }

    /** True when notifications can actually be posted right now. */
    val isAuthorized: Boolean get() = status.isAuthorized

    /**
     * Whether the system prompt has already been shown.
     *
     * Android cannot distinguish "never asked" from "asked and denied twice" without an
     * Activity, so the SDK records the fact itself. Without this, a permanently denied user
     * would be reported as [PushPermissionStatus.NOT_DETERMINED] forever.
     */
    var hasBeenAsked: Boolean
        get() = storage.getBoolean(StorageManager.KEY_PERMISSION_REQUESTED, false)
        set(value) = storage.putBoolean(StorageManager.KEY_PERMISSION_REQUESTED, value)

    /** True when a system prompt is still possible; false when only Settings can help. */
    val canRequest: Boolean
        get() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
            status == PushPermissionStatus.NOT_DETERMINED

    /**
     * Records the permission state and reports whether it changed.
     *
     * The backend needs to know when a device stops being reachable, and the OS never tells the
     * application that the user switched notifications off, so this is sampled on every
     * initialization and on every permission request.
     */
    fun recordStatusChange(): Boolean {
        val current = status.name
        val previous = storage.getString(StorageManager.KEY_LAST_PERMISSION_STATE)
        if (previous == current) return false
        storage.putString(StorageManager.KEY_LAST_PERMISSION_STATE, current)
        PushLogger.i { "Permission status: ${previous ?: "unknown"} -> $current" }
        return true
    }

    /** Intent that opens this application's notification settings, for the denied case. */
    fun notificationSettingsIntent(): Intent =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            Intent(Settings.ACTION_APP_NOTIFICATION_SETTINGS)
                .putExtra(Settings.EXTRA_APP_PACKAGE, appContext.packageName)
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        } else {
            Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
                .setData(Uri.fromParts("package", appContext.packageName, null))
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
}
